Have Init return a logger

adjackura opened the pull request:
Update README

adjackura committed:
Have Init return a logger

pkern commented on logger.go:
```diff
@@ -25,17 +25,21 @@ import (
+
 // Init sets up logging and should be called before log functions, usually in
-// the callers main(). Log functions can be called before Init(), but log
+// the callers main(). Default log functions can be called before Init(), but log
```
caller's main

pkern commented on logger.go:
```diff
@@ -58,10 +62,18 @@ func Init(name string, verbose, systemLog bool, logFile io.Writer) {
+	l.errorLog = log.New(io.MultiWriter(eLogs...), "ERROR: ", flags)
+	l.fatalLog = log.New(io.MultiWriter(eLogs...), "FATAL: ", flags)
+
+	if !initialized {
```
This global variable is now a little unfortunate, as it can race. (Which wasn't true before.) This needs to be guarded with a mutex. (Go is free to reorder the defaultLogger and initialized assignments as per the memory model as long as it doesn't change behavior within the same goroutine.)

adjackura replied:
I ended up putting the lock as a package var, multiple loggers probably wont be common but I want to make sure logging doesn't race, and some waiting shouldn't hurt logging.

pkern replied:
Given that Init() will be called infrequently I think it would suffice to just always Lock(), defer Unlock() and check against defaultLogger itself, dropping initialized entirely?

adjackura replied:
I made initialized a logger field so I can check that instead of a package var, checking the individual loggers was cumbersome

pkern commented on logger.go:
```diff
@@ -72,87 +84,139 @@ const (
+	infoLog  *log.Logger
+	errorLog *log.Logger
+	fatalLog *log.Logger
+	logLock  sync.Mutex
```
Optional: Could this sync.Mutex simply be embedded? Especially as logger is a non-exported type.

pkern commented on logger.go:
```diff
@@ -72,87 +84,139 @@ const (
 	initText := "ERROR: Logging before logger.Init."
 	switch s {
 	case sInfo:
 		if !initialized {
```
Would there be a way of implementing this using a more sensible default logger rather than accessing initialized? (Which would imply grabbing the mutex as per the above.)

pkern commented on logger.go:
```diff
@@ -72,87 +84,139 @@ const (
 	os.Exit(1)
 }
+
+// Info calls the default loggers Info.
```
These comments should all use possessive form: [...] logger's [...]

pkern commented on logger_windows.go:
```diff
@@ -38,8 +38,11 @@ func (w *writer) Write(b []byte) (int, error) {
+	// for pre-existing eventlog sources.
 	if err := eventlog.InstallAsEventCreate(src, eventlog.Info|eventlog.Error); err != nil {
-		if !strings.Contains(err.Error(), "registry key already exists") {
+		if !strings.Contains(err.Error(), "registry key already exists") && !strings.Contains(err.Error(), "Access is denied") {
```
I found where "registry key already exists" comes from and I think that library must be fixed to return a properly defined and exported error type. Do you know where "Access is denied" is emitted?

pkern reviewed (changes requested):
I'm new to GitHub code reviews, but it feels odd to me that multiple different things (like Travis & Windows fixes) are mingled into a single pull request, or at least a single Review.

adjackura committed:
Update how the defaultLogger is generated.

A defaultLogger is now assigned in package init()
Logging before Init() will now use this defaultLogger

adjackura commented:
Removed the travis.yml, will add that in a subsequent PR

pkern commented on logger.go:
```diff
@@ -27,14 +27,26 @@ import (
+const flags = log.Ldate | log.Lmicroseconds | log.Lshortfile
+
+func initialize() {
+	initText := "ERROR: Logging before logger.Init.\n"
```
const initText = "..."

pkern commented on logger.go:
```diff
@@ -25,17 +25,33 @@ import (
+}
+
+func init() {
+	initialize()
```
Do we really need this indirection? I see that you're calling initialize() in your test to test logging before Init(), but the same would be achievable if the default setup just happened during init(), I think?

adjackura replied:
That would require the tests to run in a specific order, or rewrite the initialization in the tes, I felt this was cleaner.

pkern reviewed (changes requested):
Sorry, I'm still in a remote office (NYC) and hence slightly delayed. I'll try to be quicker. But I think we're almost done here.

pkern reviewed (approved):
I think we're down to stylistic questions that we should ignore for now as they wouldn't concern the external interface. Let's land this. Would it be possible for you to squash this into two commits, one for the default change and one for the error handling? Giving LGTM ahead of time assuming that this happens. :)
